common/parallel/TreeCrawler.go

```diff
@@ -101,29 +154,117 @@ func (c *crawler) workerLoop(ctx context.Context, wg *sync.WaitGroup, workerInde
 	var err error
 	mayHaveMore := true
 	for mayHaveMore && ctx.Err() == nil {
-		mayHaveMore, err = c.processOneDirectory(ctx, workerIndex)
+		mayHaveMore, err = c.processOneDirectoryWithAutoPacer(ctx, workerIndex)
 		if err != nil {
 			c.output <- CrawlResult{err: err}
 			// output the error, but we don't necessarily stop the enumeration (e.g. it might be one unreadable dir)
 		}
 	}
 }
 
-func (c *crawler) processOneDirectory(ctx context.Context, workerIndex int) (bool, error) {
+// receiveOnTqueue receive dirs enumerated by source and append to unstartedDirs.
+func (c *crawler) receiveOnTqueue() {
+	const maxQueueDirectories = 1000 * 1000
+	// Wait for any new entries on channel.
+	for tDir := range c.tqueue {
+
+		// Lets put backpressure to source to slow down, otherwise c.unstartedDirs will keep growing and
+		// and cause memory pressure on system.
+		for len(c.unstartedDirs) > maxQueueDirectories {
+			time.Sleep(10 * time.Millisecond)
+		}
+		c.cond.L.Lock()
+		if tDir != nil {
+			c.unstartedDirs = append(c.unstartedDirs, tDir)
+		}
+		c.cond.L.Unlock()
+		c.cond.Broadcast()
+	}
+
+	// There is no way to know if channel is closed or not without interacting with channel.
+	// So, set this flag for target traverser to know source traverser done.
```

Comment:
Nit, this comment is not very clear since this function runs as part of target traverser only, it gives an impression as if this is source traverser.

	//
	// Let processOneDirectoryWithAutoPacer() (called by target traverser) know that we have flushed
	// tqueue and no new directories are expected. Once it processes everything queued to unstartedDirs
	// it may plan to exit.
	//